CORE-11065 Custom ledger queries liquibase changes

lankydan opened the pull request:
Add `custom_representation` and repurpose the `consumed` column
in the `utxo_relevant_transaction_state` table.

The `custom_representation` column is of type `JSONB` and is
therefore Postgres  specific.

`ledger-utxo-relevant-transaction-state-creation-v1.0.xml` has been 
separated to create the table. For now it has been restricted to 
creating the table only for Postgres databases, meaning that the table
will not be made for anything else. In the future, we can add the table
with the correct columns when we support more databases.

lankydan committed:
CORE-11065 Custom ledger queries liquibase changes

Add `custom` and `consumed_timestamp` to the
`utxo_relevant_transaction_state` table.

The `custom` column is of type `JSONB` and is therefore Postgres
specific.

`ledger-utxo-relevant-transaction-state-creation-v1.0.xml` has been
separated to create the table. For now it has been restricted to
creating the table only for Postgres databases, meaning that the table
will not be made for anything else. In the future, we can add the table
with the correct columns when we support more databases.

vlajos commented on data/db-schema/src/main/resources/net/corda/db/schema/vnode-vault/migration/ledger-utxo-relevant-transaction-state-creation-v1.0.xml:
```diff
@@ -0,0 +1,40 @@
+            <column name="leaf_idx" type="INT">
+                <constraints nullable="false"/>
+            </column>
+            <column name="consumed" type="BOOLEAN">
```
I guess this field could be computed from the `consumed_timestamp`.
Although I am not too sure if it makes any real differences.

blsemo replied:
Yes, I agree that this is redundant information. Having this and `consumed_timestamp` could lead to inconsistent data.

lankydan replied:
Removed the original `consumed` column and renamed `consumed_timestamp` to `consumed` to match the `created` column.

vlajos replied:
Thanks.

blsemo commented on data/db-schema/src/main/resources/net/corda/db/schema/vnode-vault/migration/ledger-utxo-relevant-transaction-state-creation-v1.0.xml:
```diff
@@ -0,0 +1,40 @@
+            <column name="consumed" type="BOOLEAN">
+                <constraints nullable="false"/>
+            </column>
+            <column name="custom" type="JSONB">
```
I am not thrilled by this name - could we have something like "custom_representation" or "custom_data" etc?

lankydan replied:
Renamed to `custom_representation`.

blsemo commented on data/db-schema/src/main/resources/net/corda/db/schema/vnode-vault/migration/ledger-utxo-creation-v1.0.xml:
```diff
@@ -217,4 +191,6 @@
                        tableName="utxo_transaction_cpk"/>
     </changeSet>
+
+    <include file="net/corda/db/schema/vnode-vault/migration/ledger-utxo-relevant-transaction-state-creation-v1.0.xml"/>
```
Good idea to split this out into a separate file!

driessamyn replied:
I don't think it makes technically a difference, but wouldn't it be more consistent to have this include in the `db.changelog-master.xml` file?
Where it is here may get a bit confusing when a new version is introduced.

lankydan committed:
CORE-11065 PR comment - Rename column

lankydan committed:
CORE-11065 Store consumed timestamp in `consumed` column

Remove the original boolean held in the `consumed` column and store the
timestamp there instead. This is because we know a state is consumed if
this timestamp is filled in, so having the separate column was
redundant.

lankydan committed:
rebased

mnesbit reviewed (approved):
OK

driessamyn commented on data/db-schema/src/main/resources/net/corda/db/schema/vnode-vault/migration/ledger-utxo-relevant-transaction-state-creation-v1.0.xml:
```diff
@@ -0,0 +1,37 @@
+                   xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-4.3.xsd">
+
+    <changeSet author="R3.Corda" id="ledger-utxo-relevant-transaction-state-creation-postgres-v1.0" dbms="postgresql">
```
For reference, where we've had to deal with DB specific types elsewhere, we have used properties.

E.g.: https://github.com/corda/corda-api/blob/5169296760cc10fbdd07ba3fee2a68227dd6adb6/data/db-schema/src/main/resources/net/corda/db/schema/config/migration/config-creation-v1.0.xml#L6

This would let you do something like
```
<property name="json-type" value="VARCHAR(MAX)" dbms="hsqldb"/>
<property name="json-type" value="JSONP" dbms="postgresql"/>

...

<column name="custom_representation" type="${json-type}">
    <constraints nullable="false"/>
</column>

```
